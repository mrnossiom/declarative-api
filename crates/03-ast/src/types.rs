use lexer::{span::Span, symbols::Symbol};

#[derive(Debug)]
pub struct Api {
	pub attrs: Vec<Attribute>,
	pub meta: Metadata,
	pub items: Vec<Item>,
	pub span: Span,
}

#[derive(Debug)]
pub struct Metadata {
	pub name: String,
	pub description: String,
	pub licence: String,
	pub urls: Vec<String>,
}

#[derive(Debug)]
pub struct Item {
	pub attrs: Vec<Attribute>,
	pub kind: ItemKind,
	pub span: Span,
	pub ident: Ident,
}

#[derive(Debug)]
pub enum ItemKind {
	Meta(Meta),

	Scope(ScopeKind),
	Path(Path),
	Model(Model),
	Headers(Headers),
	Verb(Verb),
	StatusCode(StatusCode),
}

#[derive(Debug)]
pub struct Verb {
	pub method: String,
	pub items: Vec<Item>,
}

#[derive(Debug)]
pub struct StatusCode {
	pub code: u16,
	pub items: Vec<Item>,
}

#[derive(Debug)]
pub struct Meta {
	pub fields: Vec<MetaField>,
}

#[derive(Debug)]
pub struct Headers {
	pub headers: Vec<HeaderField>,
}

#[derive(Debug)]
pub struct HeaderField {
	pub ident: Ident,
	pub attrs: Vec<Attribute>,
	pub span: Span,
}

#[derive(Debug)]
pub struct Path {
	pub path: PathKind,
	pub items: Vec<Item>,
}

#[derive(Debug)]
pub enum PathKind {
	String(Ident),
	Variable(Ident),
	Complex(Vec<Self>),
}

#[derive(Debug)]
pub struct Model {
	pub fields: Vec<ModelField>,
}
#[derive(Debug)]
pub struct ModelField {
	pub ident: Ident,
	pub ty: Type,
	pub attrs: Vec<Attribute>,
	pub span: Span,
}

#[derive(Debug)]
pub struct MetaField {
	pub ident: Ident,
	pub value: MetaFieldKind,
	pub span: Span,
}

#[derive(Debug)]
pub enum MetaFieldKind {
	Str(String),
	Bool(bool),
	Int(u64),
	Float(f64),
	Vec(Vec<Self>),
}

#[derive(Debug)]
pub struct Type(pub(crate) String);

#[derive(Debug)]
pub struct Attribute {
	pub ident: Ident,
	pub value: String,
	pub span: Span,
}

#[derive(Debug)]
pub enum ScopeKind {
	Loaded {
		items: Vec<Item>,
		/// Whether the scope was defined inline or in an external file.
		inline: bool,
		span: Span,
	},
	Unloaded,
}

#[derive(Debug)]
pub struct Ident {
	pub symbol: Symbol,
	pub span: Span,
}
