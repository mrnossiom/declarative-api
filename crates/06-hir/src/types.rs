#[derive(Debug)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

#[derive(Debug)]
pub struct ApiMetadata {
	pub name: Option<String>,
	pub version: Option<Version>,
	pub licence: Option<String>,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Response {
	pub status_code: u32,
	pub body: Vec<KeyValuePair>,
	pub headers: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Method {
	pub method: http::Method,
	pub responses: Vec<Response>,
	pub headers: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Path {
	pub name: String,
	pub scopes: Vec<Scope>,
	pub paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub headers: Vec<KeyValuePair>,
	pub parameters: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub metadata: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

pub type KeyValuePairParameter = (String, String);

#[derive(Debug, Clone)]
pub struct KeyValuePair {
	pub key: String,
	pub type_: Type,
	pub description: String,
	pub parameters: Vec<KeyValuePairParameter>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Scope {
	pub name: String,
	pub scopes: Vec<Scope>,
	pub models: Vec<Model>,
	pub paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub headers: Vec<KeyValuePair>,
	pub parameters: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Model {
	pub name: String,
	pub model_body: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct ApiData {
	pub scopes: Vec<Scope>,
}

#[derive(Debug)]
pub struct Api {
	pub metadata: ApiMetadata,
	pub data: ApiData,
}

#[derive(Debug, Clone)]
pub enum Type {
	Int,

	String,

	Model(String),
	Object(Vec<KeyValuePair>),
	List(Vec<Type>),
}
