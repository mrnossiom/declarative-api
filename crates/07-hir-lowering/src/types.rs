use std::{
	cmp::Ordering,
	collections::{BTreeMap, HashMap, HashSet},
	fmt,
	hash::Hash,
	path::Path,
};

#[derive(Debug, Clone)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

impl fmt::Display for Version {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
	}
}

impl PartialOrd for Version {
	// Required method
	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
		self.major
			.partial_cmp(&other.major)
			.and_then(|ord| match ord {
				Ordering::Equal => self.minor.partial_cmp(&other.minor),
				_ => Some(ord),
			})
			.and_then(|ord| match ord {
				Ordering::Equal => self.patch.partial_cmp(&other.patch),
				_ => Some(ord),
			})
	}
}

impl PartialEq for Version {
	fn eq(&self, other: &Self) -> bool {
		self.major == other.major && self.minor == other.minor && self.patch == other.patch
	}
}

pub trait ResolveModels {
	fn resolve_models(
		&mut self,
		scope: &ScopePath,
		intermediate_representation: IntermediateRepresentation,
	);
}

pub type ScopePath = Vec<String>;

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub struct KeyValuePairValue {
	pub type_: Type,
	pub description: String,
	pub parameters: KeyValuePairValueParameters,
	pub comment: Option<String>,
}

impl KeyValuePairValue {
	#[must_use]
	pub fn from_ast_key_value_pair_ref(
		key_value_pair: &ast::types::KeyValuePair,
	) -> (String, Self) {
		(
			key_value_pair.key.clone(),
			Self {
				type_: Type::from_ast_type(key_value_pair.type_.clone()),
				description: key_value_pair.description.clone(),
				parameters: Self::parse_parameters(&key_value_pair.parameters),
				comment: key_value_pair.comment.clone(),
			},
		)
	}
	#[must_use]
	#[allow(clippy::missing_panics_doc)]
	pub fn map_from_ast_key_value_pair_vec(
		key_value_pair_vec: &[ast::types::KeyValuePair],
	) -> BTreeMap<String, Self> {
		let mut btreemap = BTreeMap::new();
		for field in key_value_pair_vec {
			let param_btreemap = Self::parse_parameters(&field.parameters);
			assert!(
				!btreemap.contains_key(&field.key),
				"a key-value pair has the same key defined multiple times"
			);
			btreemap.insert(
				field.key.clone(),
				Self {
					type_: Type::from_ast_type(field.type_.clone()),
					description: field.description.clone(),
					parameters: param_btreemap,
					comment: field.comment.clone(),
				},
			);
		}
		btreemap
	}
	// This merges the b BTreeMap into the a BTreeMap. In case of the same keys being defined in both BTreeMap, b's values has priority.
	pub fn merge(a: &mut BTreeMap<String, Self>, b: &BTreeMap<String, Self>) {
		for (key, value) in b {
			a.insert(key.clone(), value.clone());
		}
	}
	#[must_use]
	#[allow(clippy::missing_panics_doc)]
	pub fn parse_parameters(
		parameters: &[ast::types::KeyValuePairParameter],
	) -> BTreeMap<String, String> {
		let mut param_btreemap = BTreeMap::new();
		for parameter in parameters {
			assert!(
				!param_btreemap.contains_key(&parameter.0.clone()),
				"a key-value pair has the same parameter defined multiple times"
			);
			param_btreemap.insert(parameter.0.clone(), parameter.1.clone());
		}
		param_btreemap
	}
}

impl ResolveModels for BTreeMap<String, KeyValuePairValue> {
	fn resolve_models(
		&mut self,
		scope: &ScopePath,
		intermediate_representation: IntermediateRepresentation,
	) {
		for value in self.values_mut() {
			match &mut value.type_ {
				Type::Model(model_name) => {
					let mut found = false;
					let mut current_scope = scope.clone();
					while !found {
						found = intermediate_representation
							.scopes
							.get(&current_scope)
							.expect("missing scope")
							.models
							.contains_key(model_name);
						if !found {
							assert!(
								!current_scope.is_empty(),
								"unknown reference to model \"{model_name}\""
							);
							current_scope.truncate(current_scope.len() - 1);
						}
					}
					current_scope.push(model_name.clone());
					value.type_ = Type::ResolvedModel(current_scope);
				}
				Type::Object(child_object) => {
					child_object.resolve_models(scope, intermediate_representation.clone());
				}
				_ => (),
			};
		}
	}
}

#[derive(Debug, PartialEq, Clone, Eq, Hash)]
pub struct Method {
	pub parameters: BTreeMap<String, KeyValuePairValue>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub query_params: BTreeMap<String, KeyValuePairValue>,
	pub responses: BTreeMap<u32, Response>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub struct Response {
	pub body: BTreeMap<String, KeyValuePairValue>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Eq, Hash, Default)]
pub struct Endpoint {
	pub methods: BTreeMap<String, Method>,
}

#[derive(Debug, PartialEq, Eq, Clone, Default)]
pub struct Scope {
	pub child_scopes: HashSet<String>,
	pub models: BTreeMap<String, Model>,
	pub endpoints: HashMap<String, Endpoint>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Eq, Hash, Default)]
pub struct Model {
	pub model_body: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

pub type KeyValuePairValueParameters = BTreeMap<String, String>;

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub enum Type {
	Int,

	String,

	Model(String),
	ResolvedModel(ScopePath),
	Object(BTreeMap<String, KeyValuePairValue>),
	List(Vec<Type>),
}

impl Type {
	#[must_use]
	pub fn from_ast_type(input: ast::types::Type) -> Self {
		match input {
			ast::types::Type::Int => Self::Int,
			ast::types::Type::String => Self::String,
			ast::types::Type::Model(model_name) => Self::Model(model_name),
			ast::types::Type::Object(fields) => {
				Self::Object(KeyValuePairValue::map_from_ast_key_value_pair_vec(&fields))
			}
			ast::types::Type::List(list_obj) => {
				let mut out_list_obj = vec![];
				for obj in &list_obj {
					out_list_obj.push(Self::from_ast_type(obj.clone()));
				}
				Self::List(out_list_obj)
			}
		}
	}
}

#[derive(Debug, PartialEq, Clone)]
pub struct ApiMetadata {
	pub name: String,
	pub version: Version,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct IntermediateRepresentation {
	pub metadata: ApiMetadata,
	pub scopes: BTreeMap<ScopePath, Scope>,
}

impl Default for IntermediateRepresentation {
	fn default() -> Self {
		let mut scopes_map = BTreeMap::new();
		scopes_map.insert(vec![], Scope::default());
		Self {
			metadata: ApiMetadata {
				name: String::new(),
				version: Version {
					major: 0,
					minor: 0,
					patch: 0,
				},
				urls: vec![],
				comment: None,
			},
			scopes: scopes_map,
		}
	}
}

#[derive(Debug, Clone)]
pub struct ParserVariables {
	pub current_method: String,
	pub scope_path: ScopePath,
	pub endpoint_path: Box<Path>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub parameters: BTreeMap<String, KeyValuePairValue>,
	pub query_params: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

impl Default for ParserVariables {
	fn default() -> Self {
		Self {
			current_method: String::new(),
			scope_path: ScopePath::new(),
			endpoint_path: Path::new("/").into(),
			headers: BTreeMap::new(),
			parameters: BTreeMap::new(),
			query_params: BTreeMap::new(),
			comment: None,
		}
	}
}
