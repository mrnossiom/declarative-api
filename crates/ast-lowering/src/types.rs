use std::{
	cmp::Ordering,
	collections::{BTreeMap, HashMap, HashSet},
	fmt,
	hash::Hash,
	path::Path,
	vec,
};

#[derive(Debug)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

impl fmt::Display for Version {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
	}
}

impl PartialOrd for Version {
	// Required method
	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
		self.major
			.partial_cmp(&other.major)
			.and_then(|ord| match ord {
				Ordering::Equal => self.minor.partial_cmp(&other.minor),
				_ => Some(ord),
			})
			.and_then(|ord| match ord {
				Ordering::Equal => self.patch.partial_cmp(&other.patch),
				_ => Some(ord),
			})
	}
}

impl PartialEq for Version {
	fn eq(&self, other: &Self) -> bool {
		self.major == other.major && self.minor == other.minor && self.patch == other.patch
	}
}

pub type ScopePath = Vec<String>;

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub struct KeyValuePairValue {
	pub type_: Type,
	pub description: String,
	pub parameters: BTreeMap<String, String>,
	pub comment: Option<String>,
}

impl KeyValuePairValue {
	pub fn from_ast_key_value_pair_ref(
		key_value_pair: &ast::types::KeyValuePair,
	) -> (String, Self) {
		(
			key_value_pair.key.clone(),
			Self {
				type_: Type::from_ast_type(key_value_pair.type_.clone()),
				description: key_value_pair.description.clone(),
				parameters: key_value_pair.parameters.clone(),
				comment: key_value_pair.comment.clone(),
			},
		)
	}
	pub fn map_from_ast_key_value_pair_vec(
		key_value_pair_vec: &[ast::types::KeyValuePair],
	) -> BTreeMap<String, Self> {
		let mut b_tree_map = BTreeMap::new();
		for key_value_pair in key_value_pair_vec.iter() {
			let (key, value) = Self::from_ast_key_value_pair_ref(key_value_pair);
			b_tree_map.insert(key, value);
		}
		b_tree_map
	}
}

#[derive(Debug, PartialEq, Clone, Eq, Hash)]
pub struct Method {
	pub parameters: BTreeMap<String, KeyValuePairValue>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub query_params: BTreeMap<String, KeyValuePairValue>,
	pub responses: BTreeMap<u32, Response>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub struct Response {
	pub body: BTreeMap<String, KeyValuePairValue>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Eq, Hash, Default)]
pub struct Endpoint {
	pub methods: BTreeMap<String, Method>,
}

#[derive(Debug, PartialEq, Eq, Clone, Default)]
pub struct Scope {
	pub child_scopes: HashSet<String>,
	pub models: HashSet<Model>,
	pub endpoints: HashMap<String, Endpoint>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Eq, Hash, Default)]
pub struct Model {
	pub name: String,
	pub model_body: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub enum Type {
	Int,

	String,

	Model(String),
	Object(BTreeMap<String, KeyValuePairValue>),
	List(Vec<Type>),
}

impl Type {
	pub fn from_ast_type(input: ast::types::Type) -> Self {
		match input {
			ast::types::Type::Int => Self::Int,
			ast::types::Type::String => Self::String,
			ast::types::Type::Model(model_name) => Self::Model(model_name),
			ast::types::Type::Object(fields) => {
				let mut btreemap = BTreeMap::new();
				for field in &fields {
					btreemap.insert(
						field.key.clone(),
						KeyValuePairValue {
							type_: Self::from_ast_type(field.type_.clone()),
							description: field.description.clone(),
							parameters: field.parameters.clone(),
							comment: field.comment.clone(),
						},
					);
				}
				Self::Object(btreemap)
			}
			ast::types::Type::List(list_obj) => {
				let mut out_list_obj = vec![];
				for obj in &list_obj {
					out_list_obj.push(Self::from_ast_type(obj.clone()));
				}
				Self::List(out_list_obj)
			}
		}
	}
}

#[derive(Debug, PartialEq)]
pub struct ApiMetadata {
	pub name: String,
	pub version: Version,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq)]
pub struct IntermediateRepresentation {
	pub metadata: ApiMetadata,
	pub scopes: BTreeMap<ScopePath, Scope>,
	pub models: BTreeMap<ScopePath, Model>,
}

impl Default for IntermediateRepresentation {
	fn default() -> Self {
		let mut scopes_map = BTreeMap::new();
		scopes_map.insert(vec![], Scope::default());
		Self {
			metadata: ApiMetadata {
				name: String::new(),
				version: Version {
					major: 0,
					minor: 0,
					patch: 0,
				},
				urls: vec![],
				comment: None,
			},
			scopes: scopes_map,
			models: BTreeMap::new(),
		}
	}
}

#[derive(Debug, Clone)]
pub struct ParserVariables {
	pub current_method: String,
	pub scope_path: ScopePath,
	pub endpoint_path: Box<Path>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub parameters: BTreeMap<String, KeyValuePairValue>,
	pub query_params: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

impl Default for ParserVariables {
	fn default() -> Self {
		Self {
			current_method: String::new(),
			scope_path: ScopePath::new(),
			endpoint_path: Path::new("/").into(),
			headers: BTreeMap::new(),
			parameters: BTreeMap::new(),
			query_params: BTreeMap::new(),
			comment: None,
		}
	}
}
