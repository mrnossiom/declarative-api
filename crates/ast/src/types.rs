use std::collections::BTreeMap;

#[derive(Debug)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

#[derive(Debug)]
pub struct ApiMetadata {
	pub name: Option<String>,
	pub version: Option<Version>,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Response {
	pub status_code: u32,
	pub body: Vec<KeyValuePair>,
	pub headers: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Method {
	pub name: http::Method,
	pub responses: Vec<Response>,
	pub headers: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Path {
	pub name: String,
	pub child_scopes: Vec<Scope>,
	pub child_paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub headers: Vec<KeyValuePair>,
	pub parameters: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub metadata: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug, Clone)]
pub struct KeyValuePair {
	pub key: String,
	pub type_: Type,
	pub description: String,
	pub parameters: BTreeMap<String, String>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Scope {
	pub name: String,
	pub child_scopes: Vec<Scope>,
	pub child_models: Vec<Model>,
	pub child_paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub headers: Vec<KeyValuePair>,
	pub parameters: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Model {
	pub name: String,
	pub model_body: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct ApiData {
	pub child_scopes: Vec<Scope>,
}

#[derive(Debug)]
pub struct Api {
	pub metadata: ApiMetadata,
	pub data: ApiData,
}

#[derive(Debug, Clone)]
pub enum Type {
	Int,

	String,

	Model(String),
	Object(Box<KeyValuePair>),
}
