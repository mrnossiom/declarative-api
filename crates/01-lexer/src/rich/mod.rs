mod enricher;
mod errors;
mod token;

pub use enricher::Enricher;
pub use token::*;
