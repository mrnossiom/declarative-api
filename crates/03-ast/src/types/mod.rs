use crate::P;

mod attr;
mod expr;
mod item;
mod node;
pub use attr::*;
pub use expr::*;
pub use item::*;
pub use node::*;
use session::Span;
use thin_vec::ThinVec;

#[derive(Debug, Clone)]
pub struct Api {
	pub attrs: AttrVec,
	pub items: ThinVec<P<Item>>,

	pub id: NodeId,
	pub span: Span,
}

// TODO: nope, not a String
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Type(pub String);
