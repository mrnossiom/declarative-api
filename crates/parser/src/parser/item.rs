use crate::{error::InvalidVerb, PResult, Parser};
use dapic_ast::types::{
	Api, AttrVec, Auth, Body, Enum, Headers, Item, ItemKind, Metadata, Model, NodeId, Params,
	PathItem, PathKind, Query, ScopeKind, StatusCode, Verb, P,
};
use dapic_lexer::rich::{Delimiter, OpKind, TokenKind};
use dapic_session::{
	symbols::{kw, remarkable},
	Ident,
};
use thin_vec::{thin_vec, ThinVec};

impl<'a> Parser<'a> {
	#[tracing::instrument(level = "DEBUG", skip(self))]
	pub fn parse_root(&mut self) -> PResult<Api> {
		let lo = self.token.span;

		let attrs = self.parse_inner_attrs()?;

		let mut items = ThinVec::default();
		items.push(self.parse_metadata()?);
		items.extend(self.parse_scope_content(None)?);

		Ok(Api {
			attrs,
			items,
			id: NodeId::ROOT,
			span: self.span(lo),
		})
	}

	#[tracing::instrument(level = "DEBUG", skip(self, attrs))]
	fn parse_scope(&mut self, attrs: &mut AttrVec) -> PResult<(Ident, ScopeKind)> {
		self.expect_keyword(kw::Scope)?;

		let ident = self.parse_ident()?;

		let scope = if self.eat(&TokenKind::Semi) {
			ScopeKind::Unloaded
		} else {
			let lo = self.token.span;
			let items = self.expect_braced(|p| p.parse_scope_content(Some(attrs)))?;

			ScopeKind::Loaded {
				items,
				inline: true,
				span: self.span(lo),
			}
		};

		Ok((ident, scope))
	}

	#[tracing::instrument(level = "DEBUG", skip(self, attrs))]
	pub fn parse_scope_content(
		&mut self,
		attrs: Option<&mut AttrVec>,
	) -> PResult<ThinVec<P<Item>>> {
		if let Some(attrs) = attrs {
			attrs.extend(self.parse_inner_attrs()?);
		}

		self.parse_items()
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_metadata(&mut self) -> PResult<P<Item>> {
		let lo = self.token.span;

		if !self.eat_keyword(kw::Meta) {
			todo!("react accordingly")
		}

		let fields = self.expect_braced(Self::parse_property_defs)?;

		let meta = Metadata { fields };

		Ok(Self::make_item(
			ThinVec::default(),
			ItemKind::Meta(meta),
			None,
			self.span(lo),
		))
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_items(&mut self) -> PResult<ThinVec<P<Item>>> {
		let mut items = ThinVec::default();
		while let Some(item) = self.parse_item()? {
			items.push(item);
		}
		Ok(items)
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_item(&mut self) -> PResult<Option<P<Item>>> {
		let mut attrs = self.parse_outer_attrs()?;

		let lo = self.token.span;

		// TODO: check keywords with case insensitive

		let (ident, kind) = if self.check_keyword(kw::Scope) {
			// `scope <ident> { <items> }`
			let (ident, kind) = self.parse_scope(&mut attrs)?;
			(Some(ident), ItemKind::Scope(kind))
		} else if self.check_keyword(kw::Path) {
			// `path <path> { <items> }`
			let path = self.parse_path_item()?;
			(None, ItemKind::Path(path))
		} else if self.check_keyword(kw::Meta) {
			// `meta { <properties> }`
			// TODO: change or emit warn about misplaced metadata, should go in passes
			let metadata = self.parse_metadata()?.kind.clone();
			(None, metadata)
		} else if self.check_keyword(kw::Headers) {
			// `headers { <fields> }`
			let headers = self.parse_headers()?;
			(None, ItemKind::Headers(headers))
		} else if self.check_keyword(kw::Query) {
			// `query { <fields> }`
			let query = self.parse_query()?;
			(None, ItemKind::Query(query))
		} else if self.check_keyword(kw::Code) {
			// `code <code lit> { <items> }`
			let item = self.parse_code_item()?;
			(None, ItemKind::StatusCode(item))
		} else if self.check_keyword(kw::Model) {
			// `model <ident> { <def_fields> }`
			let (ident, item) = self.parse_model()?;
			(Some(ident), ItemKind::Model(item))
		} else if self.check_keyword(kw::Enum) {
			// `enum <ident> { <def_fields> }`
			let (ident, item) = self.parse_enum()?;
			(Some(ident), ItemKind::Enum(item))
		} else if self.check_keyword(kw::Auth) {
			// `auth <ident> { <auth_fields> }`
			let (ident, item) = self.parse_auth()?;
			(Some(ident), ItemKind::Auth(item))
		} else if self.check_keyword(kw::Verb) {
			// `verb <http_verb_ident> { <items> }`
			let (ident, verb) = self.parse_verb()?;
			(Some(ident), ItemKind::Verb(verb))
		} else if self.check_keyword(kw::Body) {
			// `body <ty>`
			let body = self.parse_body()?;
			(None, ItemKind::Body(body))
		} else if self.check_keyword(kw::Params) {
			// `params { <field_defs> }`
			let params = self.parse_params()?;
			(None, ItemKind::Params(params))
		} else {
			if attrs.is_empty() {
				return Ok(None);
			}

			todo!("error if attributes are parsed but there is not item to attach them")
		};

		Ok(Some(Self::make_item(attrs, kind, ident, self.span(lo))))
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_path_item(&mut self) -> PResult<PathItem> {
		self.expect_keyword(kw::Path)?;
		let kind = self.parse_path_item_kind()?;
		let items = self.expect_braced(Self::parse_items)?;
		Ok(PathItem { kind, items })
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_path_item_kind(&mut self) -> PResult<PathKind> {
		let kind = if self.eat(&TokenKind::OpenDelim(Delimiter::Brace)) {
			let ident = self.parse_ident()?;
			self.expect(&TokenKind::CloseDelim(Delimiter::Brace))?;
			PathKind::Variable(ident)
		} else if self.eat(&TokenKind::Dot) {
			PathKind::Current
		} else {
			let ident = self.parse_ident()?;

			PathKind::Simple(ident)
		};

		if self.eat(&TokenKind::Op(OpKind::Slash)) {
			let mut parts = thin_vec![kind];

			match self.parse_path_item_kind()? {
				PathKind::Complex(vec) => parts.extend(vec),
				pk @ (PathKind::Simple(..) | PathKind::Variable(..) | PathKind::Current) => {
					parts.push(pk);
				}
			}

			Ok(PathKind::Complex(parts))
		} else {
			Ok(kind)
		}
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_headers(&mut self) -> PResult<Headers> {
		self.expect_keyword(kw::Headers)?;
		let headers = self.expect_braced(Self::parse_field_defs)?;
		Ok(Headers { headers })
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_query(&mut self) -> PResult<Query> {
		self.expect_keyword(kw::Query)?;
		let fields = self.expect_braced(Self::parse_field_defs)?;
		Ok(Query { fields })
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_code_item(&mut self) -> PResult<StatusCode> {
		self.expect_keyword(kw::Code)?;
		let code = self.parse_expr()?;
		let items = self.expect_braced(Self::parse_items)?;
		Ok(StatusCode { code, items })
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_verb(&mut self) -> PResult<(Ident, Verb)> {
		self.expect_keyword(kw::Verb)?;
		let method = self.parse_ident()?;
		let items = self.expect_braced(Self::parse_items)?;

		{
			// TODO: this is an experimental lint
			// this should be in a lint passes
			// also custom verbs are allowed

			use remarkable::{Connect, Delete, Get, Head, Options, Post, Put, Trace};
			if ![Connect, Delete, Get, Head, Options, Post, Put, Trace].contains(&method.symbol) {
				return Err(InvalidVerb { found: method }.into());
			}
		}

		Ok((method, Verb { method, items }))
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_model(&mut self) -> PResult<(Ident, Model)> {
		self.expect_keyword(kw::Model)?;
		let name = self.parse_ident()?;
		let fields = self.expect_braced(Self::parse_field_defs)?;
		Ok((name, Model { fields }))
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_enum(&mut self) -> PResult<(Ident, Enum)> {
		self.expect_keyword(kw::Enum)?;
		let name = self.parse_ident()?;
		let variants = self.expect_braced(Self::parse_property_defs)?;
		Ok((name, Enum { variants }))
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_auth(&mut self) -> PResult<(Ident, Auth)> {
		self.expect_keyword(kw::Auth)?;
		let auth_name = self.parse_ident()?;

		let kind = if self.eat(&TokenKind::Semi) {
			// `auth BasicAuth;`
			Auth::Use
		} else {
			// `auth BasicAuth { <field_defs> }`

			// TODO: define `auth` syntax and parse it
			let _fields = self.expect_braced(Self::parse_field_defs)?;

			Auth::Define {}
		};

		Ok((auth_name, kind))
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_body(&mut self) -> PResult<Body> {
		self.expect_keyword(kw::Body)?;
		let ty = self.parse_ty()?;
		Ok(Body { ty })
	}

	#[tracing::instrument(level = "DEBUG", skip(self))]
	fn parse_params(&mut self) -> PResult<Params> {
		self.expect_keyword(kw::Params)?;
		let properties = self.expect_braced(Self::parse_field_defs)?;
		Ok(Params { properties })
	}
}

#[cfg(test)]
mod tests {
	use crate::assert_tokenize;

	assert_tokenize!("simple", parse_path_item_kind, "var");
	assert_tokenize!("variable", parse_path_item_kind, "{var}");
	assert_tokenize!("complex", parse_path_item_kind, "var1/{var2}");
	assert_tokenize!("long_complex", parse_path_item_kind, "var1/{var2}/{var3}");
}
