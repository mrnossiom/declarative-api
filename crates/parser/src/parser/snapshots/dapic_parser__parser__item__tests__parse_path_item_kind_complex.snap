---
source: crates/parser/src/parser/item.rs
expression: p.parse_path_item_kind()?
---
Complex(
    [
        Simple(
            Ident {
                symbol: Symbol(29, "var1"),
                span: Span {
                    start: BytePos(
                        0,
                    ),
                    end: BytePos(
                        4,
                    ),
                },
            },
        ),
        Variable(
            Ident {
                symbol: Symbol(30, "var2"),
                span: Span {
                    start: BytePos(
                        6,
                    ),
                    end: BytePos(
                        10,
                    ),
                },
            },
        ),
    ],
)
